### PR TITLE
Revpi 2545/investigate high cyclus latency

### PR DESCRIPTION
**Pull request by linosanfilippo-kunbus**
This is base on [linosanfilippo-kunbus:revpi-2545/use_serdev_highlevel_functions](https://github.com/linosanfilippo-kunbus/linux/tree/revpi-2545/use_serdev_highlevel_functions). For review only take the last 2 commits into account until the other PR is merged.

### CODE REVIEW

**Comment by l1k** [comment]
Wow, good catch!

**Comment by l1k** [comment]
<blockquote>This is preferred to fix this issue by waking up uninterruptible tasks, since waiting uninterruptibly is almost always
a bad thing to do.</blockquote>

My understanding is that interruptible tasks are ones that user space interacts with. E.g. if user space hits ^C, a running system call is interrupted. However the (cyclic) PiBridge communication is completely obscured from user space. User space never directly interacts with it. All ioctls that perform PiBridge communication (such as `KB_DIO_RESET_COUNTER`) do so indirectly by writing the data to be sent to `piCore_g.pendingUserTel`. The `piIoThread()` independently takes care of sending the data and receiving the response.

Do you envision future use cases where user space will directly perform PiBridge communication?

Otherwise this could be solved by simply dropping the `_interruptible` from `wake_up_interruptible()`, no?

File: drivers/tty/serdev/pibridge.c [hunk 1]
```diff
@@ -87,16 +87,18 @@ static int pibridge_discard_timeout(u16 len, u16 timeout)
...
-	if (kfifo_len(&pi->read_fifo) >= len)
-		ret = 0;
-	else
-		ret = -1;
-	kfifo_reset(&pi->read_fifo);
-	mutex_unlock(&pi->lock);
+	ret = wait_event_interruptible_timeout(pi->read_queue,
+					       kfifo_len(&pi->read_fifo) >= len,
+					       msecs_to_jiffies(timeout));
+	if (ret != -ERESTARTSYS) {
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c]: Why not

```
if (ret == -ERESTARTSYS)
	return ret;
```

? Avoids the need to indent all those lines. At least that's the pattern I've seen used most of the time in the kernel.

File: drivers/tty/serdev/pibridge.c [hunk 2]
```diff
@@ -284,21 +301,32 @@ int pibridge_req_gate_tmt(u8 dst, u16 cmd, u8 *snd_buf, u8 snd_len,
 	}
 
 	if (to_discard) {
+		int ret;
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c]: Huh? Looks like you already added another one further up which is scoped to the function?

File: drivers/tty/serdev/pibridge.c [hunk 3]
```diff
@@ -423,22 +460,33 @@ int pibridge_req_io(u8 addr, u8 cmd, u8 *snd_buf, u8 snd_len, u8 *rcv_buf,
...
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c]: Same here.

File: drivers/tty/serdev/pibridge.c [hunk 4]
```diff
@@ -260,8 +270,11 @@ int pibridge_req_gate_tmt(u8 dst, u16 cmd, u8 *snd_buf, u8 snd_len,
 	if (dst == PIBRIDGE_BC_ADDR)
 		return 0;
 
-	if (pibridge_recv_timeout((u8 *) &pkthdr, sizeof(pkthdr), tmt) !=
-	    sizeof(pkthdr)) {
+	ret = pibridge_recv_timeout((u8 *) &pkthdr, sizeof(pkthdr), tmt);
+	if (ret == -ERESTARTSYS)
+		return ret;
+
+	if (ret) {
```
Comment (l1k) [inline drivers/tty/serdev/pibridge.c]: Independently of the problem you seek to solve, the `ret = ...; if (ret) ...` pattern is IMHO easier to read and closer to kernel coding style than having a function in an if-condition which has significant side effects.

**Comment by linosanfilippo-kunbus** [comment]
> My understanding is that interruptible tasks are ones that user space interacts with. E.g. if user space hits ^C, a running system call is interrupted. However the (cyclic) PiBridge communication is completely obscured from user space. User space never directly interacts with it. All ioctls that perform PiBridge communication (such as KB_DIO_RESET_COUNTER) do so indirectly by writing the data to be sent to piCore_g.pendingUserTel. The piIoThread() independently takes care of sending the data and receiving the response.
> 

"Interruptible" or "Uninterruptible" does not characterize the task itself but is a mode in which the task is suspended (so in Linux there are actually two different sleep states a task can be in). In our case we wait "uninterruptibly" which means that for the wait time all signals are ignored no matter if sent from userspace or kernel. This also includes the TERM and KILL signal sent to processes to (forcedly) terminate when the system is shut down. 

> Do you envision future use cases where user space will directly perform PiBridge communication?
> 
> Otherwise this could be solved by simply dropping the _interruptible from wake_up_interruptible(), no?

I can remember that due to the behavior concerning signals waiting uninterruptibly is not well received by the kernel folks. Thats the reason why I used the interruptible wait version. However we do the uninterruptible sleep with a timeout and so it might not be that harmful in our case.  After I thought about this again I think it might be better to keep the waiting as it is for now and indeed only adjust the wakeup. Its just too much code change involved to handle the ERESTARTSYS case and its beyond the issue we want to fix here.  So I will change the pull request accordingly and drop the second commit for now.

**Review by l1k — APPROVED** [review]

**Comment by l1k** [comment]
Ugh, can't merge this without a rebase.

**Pull request merged**